libsolidity/ast/ASTUtils.cpp

```diff
@@ -19,12 +19,35 @@
 #include <libsolidity/ast/AST.h>
 #include <libsolidity/ast/ASTUtils.h>
 
+#include <libsolutil/Algorithms.h>
+
 namespace solidity::frontend
 {
 
+bool isConstantVariableRecursive(VariableDeclaration const& _varDecl)
```

Comment:
Do we really have to duplicate the logic? Can't the check from the ``PostTypeChecker`` be moved to an earlier time instead?

Reply:
``ContractLevelChecker`` maybe? The check in the ``PostTypeChecker`` only requires the ``DeclarationTypeChecker`` to already have run, right? And one can argue that it's "Contract Level", i.e. for each contract look at its state variables...
But not sure, maybe it's more complicated than that.

Reply:
Maybe a simpler method would be to make ``rootConstVariableDeclaration`` in ``ASTUtils.cpp`` robust against cycles and return ``nullptr`` in the cycle case...

Reply:
> Can't the check from the PostTypeChecker be moved to an earlier time instead?

Maybe. It's not entirely trivial to move it, since that check runs on the entire AST and probably needs to be run after `TypeChecker`.

Anyway having the assert inside `rootConstVariableDeclaration` makes sense, because it's called from `ContractCompiler.cpp` and `IRGeneratorForStatements.cpp`.

Reply:
Not an assert - that would just ICE in these cases - just returning ``nullptr`` should be fine, though, shouldn't it? That can happen anyways and is consistent (there's not always a root variable declaration and a cycle is one such case)

Reply:
It'd probably error out with ``Only direct number constants and references to such constants are supported by inline assembly.`` that way, but I'd say that's ok...

Reply:
> return nullptr in the cycle case...

I thought about this, however we'll still have to distinguish between non-direct-constants and cyclic constants (both will return nullptr.) And the error message won't be so useful.

Reply:
> it'd probably error out with Only direct number constants and references to such constants are supported by inline assembly.

But we would miss the error about cyclic constant, since this check is done after TypeChecker and will never be done. I think this would be confusing.

Reply:
Hm, yeah, not sure... I think it'd be fine - it's true that the cyclic constant doesn't refer to a number in the end :-).
On the other hand making ``rootConstVariableDeclaration`` robust against cycles is not all that different from checking for cycles before calling it in the type checker...

What we'd actually need is a ``PreTypeChecker`` that works like the ``PostTypeChecker``, just before the ``TypeChecker`` (actual type checking is not required for the constant detection in ``PostTypeChecker`` - it only really needs ``DeclarationTypeChecker``...)... but that'd maybe be overkill :-)...

Well, I'd also be fine with just leaving this as is...

Reply:
I think it's a good idea to introduce a new component.

Reply:
@chriseth You mean a component like `PreTypeChecker` that @ekpyron mentioned?

Reply:
What other steps do you envision in that file? Maybe better to have a dedicated `CircularRefrencesChecker` step/file?

Reply:
Just this for now. We may find another crash and then we can move more things there.

Reply:
Doing `ConstStateVarCircularReferenceChecker` before TypeChecking won't work for:

```
    ==== Source: a ====
    import "b";
    uint constant c = d;
    ==== Source: b ====
    import "a" as M;
    uint constant b = M.c;
    uint constant d = b;
    contract C {
        uint constant a = b;
    }
```

The `MemberAccess.annotation().referencedDeclaration()` is set during TypeChecker. So it would miss this test case.

Can `TypeChecker::visit(MemberAccess const& _memberAccess)` be moved further up in the compiler stack, or is there a reason it's done in `TypeChecker`?

Reply:
You are right, then we can only do it in the TypeChecker :(